lib/logstash/codecs/cef.rb

```diff
@@ -58,28 +98,115 @@ def decode(data)
   end
 
   public
-  def encode(data)
+  def encode(event)
     # "CEF:0|Elasticsearch|Logstash|1.0|Signature|Name|Sev|"
 
-    # TODO: Need to check that fields are set!
+    vendor = sanitize_header_field(event.sprintf(@vendor))
+    vendor = self.class.get_config["vendor"][:default] if vendor == ""
+
+    product = sanitize_header_field(event.sprintf(@product))
+    product = self.class.get_config["product"][:default] if product == ""
+
+    version = sanitize_header_field(event.sprintf(@version))
+    version = self.class.get_config["version"][:default] if version == ""
+
+    signature = sanitize_header_field(event.sprintf(@signature))
+    signature = self.class.get_config["signature"][:default] if signature == ""
+
+    name = sanitize_header_field(event.sprintf(@name))
+    name = self.class.get_config["name"][:default] if name == ""
+
+    # :sev is deprecated and therefore only considered if :severity equals the default setting or is invalid
+    severity = sanitize_severity(event, @severity)
+    if severity == self.class.get_config["severity"][:default]
+      # Use deprecated setting sev
+      severity = sanitize_severity(event, @sev)
+    end
 
-    # Signature, Name, and Sev should be set in the config, with ref to fields
     # Should also probably set the fields sent
-    header = ["CEF:0", "Elasticsearch", "Logstash", "1.0", @signature, @name, @sev].join("|")
-    values = @fields.map {|name| get_value(name, data)}.join(" ")
-    # values = values.map {|k,v| "#{k}=#{v}"}.join(" ")
-    @on_event.call(header + " " + values + "\n")
+    header = ["CEF:0", vendor, product, version, signature, name, severity].join("|")
+    values = @fields.map {|fieldname| get_value(fieldname, event)}.compact.join(" ") unless fields.nil?
+
+    @on_event.call(event, header + "|" + values.to_s + "\n")
```

Comment:
for performance reasons, I recommend using 

```ruby
# string interpolation
@on_event.call(event, "#{header}|#{values}\n")
# instead of concatenation
@on_event.call(event, header + "|" + values.to_s + "\n")
```

the first version using interpolation creates less strings :)